### PR TITLE
fuzzy

### PR DESCRIPTION
**Pull request by Iapetus-11**
- Bot now attempts to suggest commands when a command is not found
- Tag invokes are now lenient and use fuzzy searching
- New tag search endpoint and command

image
image
image

### CODE REVIEW

File: ClemBot.Api/ClemBot.Api.Core/Features/Tags/Bot/Search.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public record Handler(ClemBotContext _context) : IRequestHandler<Command, IQueryResult<Model>>
+    {
+        public async Task<IQueryResult<Model>> Handle(Command request, CancellationToken cancellationToken)
+        {
+            var tags = await _context.Tags
+                .Where(t => t.GuildId == request.GuildId)
+                .Select(t => new {t, Similarity=EF.Functions.TrigramsSimilarity(t.Name, request.Query)})
+                .Where(e => e.Similarity > 0.1)
```
Comment (Jay-Madden) [inline ClemBot.Api/ClemBot.Api.Core/Features/Tags/Bot/Search.cs:58]: pull this out as a constant

File: ClemBot.Api/ClemBot.Api.Core/Features/Tags/Details.cs [hunk 2]
```diff
@@ -37,10 +37,19 @@ public record QueryHandler(ClemBotContext _context) : IRequestHandler<Query, IQu
...
-                .Where(g => g.GuildId == request.GuildId && g.Name == request.Name)
-                .Include(y => y.TagUses)
-                .FirstOrDefaultAsync();
+                .Where(t => t.GuildId == request.GuildId)
+                .Select(t => new {t, Similarity = EF.Functions.TrigramsSimilarity(t.Name, request.Name)})
+                .Where(t => t.Similarity > 0.4)
+                .OrderByDescending(t => t.Similarity)
+                .Select(t => t.t)
+                .Include(t => t.TagUses)
+                .FirstOrDefaultAsync(cancellationToken);
```
Comment (Jay-Madden) [inline ClemBot.Api/ClemBot.Api.Core/Features/Tags/Details.cs:52]: we dont want to run the trigram search on a details call, only on an actual search

**Review by Jay-Madden — CHANGES_REQUESTED** [review]

File: ClemBot.Api/ClemBot.Api.Core/Features/Tags/Details.cs (hunk 2 above)
Comment (Iapetus-11) [reply ClemBot.Api/ClemBot.Api.Core/Features/Tags/Details.cs:52]: This endpoint is the one called in the tags service for each regex match. Should I make a new endpoint or change the invoke endpoint?

Comment (Jay-Madden) [reply ClemBot.Api/ClemBot.Api.Core/Features/Tags/Details.cs:52]: We dont want to fuzzy match on inline syntax, if the tag doesnt exist we assume the user wasnt attempting to invoke a tag

File: ClemBot.Bot/bot/cogs/tags_cog.py [hunk 3]
```diff
@@ -180,6 +181,24 @@ async def info(self, ctx, name: str):
...
+    async def search(self, ctx: commands.Context, query: str):
+        tags = await self.bot.tag_route.search_tags(ctx.guild.id, query)
+
+        embed = discord.Embed(title=f':information_source:  Search Results', color=Colors.ClemsonOrange)
+        embed.set_footer(text=self.get_full_name(ctx.author), icon_url=ctx.author.display_avatar.url)
+
+        if tags:
+            embed.description = '\n\n'.join([f'**{i+1}.** `{tag.name}`' for i, tag in enumerate(tags)])
+        else:
+            embed.description = "No results found... :/"
```
Comment (Jay-Madden) [inline ClemBot.Bot/bot/cogs/tags_cog.py:197]:
```suggestion
            embed.description = 'No results found... :/'
```

**Review by Jay-Madden — APPROVED** [review]

**Pull request merged**